overlord/snapshotstate/backend/backend.go

```diff
@@ -157,6 +157,63 @@ func Filename(snapshot *client.Snapshot) string {
 	return filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_%s_%s_%s.zip", snapshot.SetID, snapshot.Snap, snapshot.Version, snapshot.Revision))
 }
 
+// EstimateSnapshotSize calculates estimated size of the snapshot.
+// XXX: this doesn't take users into account yet, because it is used only
+// for automatic snapshots at the moment.
+func EstimateSnapshotSize(si *snap.Info) (int64, error) {
+	var total int64
+	calculateSize := func(path string, finfo os.FileInfo, err error) error {
+		if finfo.Mode().IsRegular() {
+			total += finfo.Size()
+		}
+		return err
+	}
+
+	// helper function for visiting revisionDir and common dir that is
+	// at the same level as revisionDir.
+	visitDataDir := func(revisionDir string) error {
+		parent := filepath.Dir(revisionDir)
+		exists, isDir, err := osutil.DirExists(parent)
+		if err != nil {
+			return err
+		}
+		if exists && !isDir {
+			return nil
+		}
+		if !exists {
+			return nil
+		}
```

Comment:
Maybe combine these two tests as `if !(exists && isDir)`? This would also be the same pattern used below in the next tests.